src/main/java/org/jenkinsci/plugins/workflow/log/DelayBufferedOutputStream.java

```diff
@@ -54,6 +58,45 @@ private Tuning() {}
         static final Tuning DEFAULT = new Tuning();
     }
 
+    /**
+     * The interesting state of the buffered stream, kept as a separate object so that {@link FlushRef} can hold on to it.
+     */
+    private static final class Buffer {
+
+        final OutputStream out;
+        private final byte[] dat;
+        private int pos;
+
+        Buffer(OutputStream out, int size) {
+            this.out = out;
+            dat = new byte[size];
+        }
+
+        synchronized void drain() throws IOException {
+            if (pos == 0) {
+                return;
+            }
+            out.write(dat, 0, pos);
+            pos = 0;
+        }
+
+        void write(int b) throws IOException {
+            assert Thread.holdsLock(this);
+            if (pos == dat.length) {
+                drain();
+            }
+            dat[pos++] = (byte) b;
+        }
+
+        synchronized void write(byte[] b, int off, int len) throws IOException {
+            for (int i = 0; i < len; i++) {
+                write(b[off + i]);
```

Comment:
Urgh, this is not very efficient, better to support writing an array chunk at a time (up to the buffer capacity, anyway)

Reply:
The OpenJDK implementation does something along those lines. We cannot copy it. I chose to write the simplest possible code that could work so it is efficiently reviewable.

Reply:
More specifically, the OpenJDK implementation has an optimization for the case that `write(byte[], int, int)` is called with a `len` which is greater than the buffer size (or maybe it is the currently open space in the buffer). Since our buffer size is pretty big by default, this is unlikely to matter much. The normal mode of operation is that the buffer fills up and then is written in one chunk.

Now copying one in-memory array to another bytewise is admittedly not terribly efficient in the case of gigantic arrays (again the OpenJDK implementation has an optimization for this which we cannot legally copy), but this has no effect on I/O, and would have made the implementation more complex and harder to review.

Reply:
The problem is overheads from many small single-byte function calls vs. doing a single large-block write (and if it exceeds the buffer size we should probably write directly to the output).   It is not needed to do native-code work -- `System.arrayCopy` already does that for us, and reputedly they use hand-tuned assembly for that (benchmarks I've done in the past bear that out). 

See this logic from `BufferedInputStream`:

```java
public synchronized void write(byte b[], int off, int len) throws IOException {
        if (len >= buf.length) {
            /* If the request length exceeds the size of the output buffer,
               flush the output buffer and then write the data directly.
               In this way buffered streams will cascade harmlessly. */
            flushBuffer();
            out.write(b, off, len);
            return;
        }
        if (len > buf.length - count) {
            flushBuffer();
        }
        System.arraycopy(b, off, buf, count, len);
        count += len;
    }
```

Reply:
I.E. this isn't a problem at the *disk* level because we're going to be behind one or more levels of internal buffering (here) or at the OS level.   But given these logging methods are going to be invoked a *lot* it makes sense to keep the overhead from this operation low.

Reply:
Yes again I am well aware of the optimizations in `BufferedOutputStream`, which I deliberately did not attempt to copy.